xapian-applications/omega/clickmodel/simplifieddbn.cc

```diff
@@ -0,0 +1,219 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+
+#include "simplifieddbn.h"
+
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+using namespace std;
+
+#define QID 0
+#define DOCIDS 1
+#define CLICKS 2
+
+string
+SimplifiedDBN::name()
+{
+    return "SimplifiedDBN";
+}
+
+static vector<string>
+get_docid_list(const vector<string> &session)
+{
+    vector<string> docids;
+    string docid;
+    for (size_t j = 0; j <= session[DOCIDS].length(); ++j) {
+	char ch = session[DOCIDS][j];
+	if (ch != ',' && ch != '\0') {
+	    docid += ch;
+	} else {
+	    docids.push_back(docid);
+	    docid.clear();
+	}
+    }
+    return docids;
+}
+
+static vector<int>
+get_click_list(const vector<string> &session)
+{
+    vector<int> clicks;
+    string clickstring, clickcount;
+    int click = 0;
+
+    for (size_t j = 0; j <= session[CLICKS].length(); ++j) {
+	char ch = session[CLICKS][j];
+	if (ch != ',' && ch != '\0') {
+	    // Get clickstring of the form "docid:click_count".
+	    clickstring += session[CLICKS][j];
+
+	    // Get click count string.
+	    size_t delimiter_pos = clickstring.find(':');
+	    clickcount = clickstring.substr(delimiter_pos + 1);
+
+	    // Convert click count string to an integer.
+	    stringstream ss(clickcount);
+	    ss >> click;
+	} else {
+	    clicks.push_back(click);
+	    clickstring.clear();
+	}
+    }
+    return clicks;
+}
+
+vector<vector<string>>
+SimplifiedDBN::build_sessions(const string &logfile)
+{
+    ifstream file;
+    file.open(logfile, ios::in);
+
+    if (!file) {
+	throw runtime_error("ERROR: Specified file does not exist.");
+    }
+
+    string line;
+
+    // Skip the first line.
+    getline(file, line);
+
+    vector<vector<string>> sessions;
+    vector<string> session_entry;
+
+    // Start reading file from the second line.
+    while (getline(file, line)) {
+
+	istringstream ss(line);
+
+	vector<string> row_data;
+
+	while (ss >> std::ws) {
+	    string column_element;
+	    if (ss.peek() == '"') {
+		int pos = ss.tellg();
+		ss.seekg(pos + 1);
+		char ch;
+		while (ss.get(ch)) {
+		    if (ch == '"')
+			break;
+		    column_element += ch;
+		}
+	    } else {
+		if (ss.peek() == ',') {
+		    int pos = ss.tellg();
+		    ss.seekg(pos + 1);
+		}
+		getline(ss, column_element, ',');
+	    }
+	    row_data.push_back(column_element);
+	}
+
+	string qid = row_data[0];
+	string query = row_data[1];
+	string docids = row_data[2];
+	string clicks = row_data[4];
+
+	session_entry.push_back(qid);
+	session_entry.push_back(docids);
+	session_entry.push_back(clicks);
+
+	sessions.push_back(session_entry);
+
+	session_entry.clear();
+    }
+
+    file.close();
+    return sessions;
+}
+
+void
+SimplifiedDBN::train(vector<vector<string>> &sessions)
+{
+    map<string, map<string, map<char, double[PARAM_COUNT_]>>> url_rel_fractions;
```

Comment:
Not sure if that can be done with enum that we have because here `a` and `s` chars need to be mapped to two different values which is why an array of size 2 is required here and it's not possible to declare arrays as enum elements. What do you suggest?